astropy/io/fits/_utils.pyx

```diff
@@ -0,0 +1,57 @@
+from collections import OrderedDict
+
+cdef Py_ssize_t BLOCK_SIZE = 2880  # the FITS block size
+cdef Py_ssize_t CARD_LENGTH = 80
+cdef str VALUE_INDICATOR = '= '  # The standard FITS value indicator
+cdef str END_CARD = 'END' + ' ' * 77
+
+
+def parse_header(fileobj):
+    """The main method allowing to parse quickly a FITS header."""
+
+    cards = OrderedDict()
+    read_blocks = []
+    cdef int found_end = 0
+    cdef bytes block
+    cdef str header_str, block_str, card_image, keyword
+    cdef Py_ssize_t idx, end_idx, sep_idx
+
+    while True:
+        # iterate on blocks
+        block = fileobj.read(BLOCK_SIZE)
+        if not block or len(block) < BLOCK_SIZE:
+            # header looks incorrect, raising exception to fall back to
+            # the full Header parsing
+            raise Exception
+
+        block_str = block.decode('ascii')
```

Comment:
Aren't headers supposed to be ASCII?

Reply:
Yes, they are supposed to. But sometimes they contain unicode chars, which will fail here, and fallback to Header, which then use [`decode_ascii`](https://github.com/astropy/astropy/blob/18515587412b6606cd3549baafb29583d91ba6a1/astropy/io/fits/util.py#L276). 
One other option, that I tried, would be to keep working on bytes here. And store the header string as bytes, which would save some memory and avoid decoding headers that are not used later. But there are some complications, like the need to check for padding bytes, and the gain was null... but I will probably try again this later.